### PR TITLE
feat(route): Telegram Channels (with video and files)

### PR DESCRIPTION
**Pull request by synchrone**
## 该 PR 相关 Issue / Involved Issue

Related to #8031

## 路由地址示例 / Example for the Proposed Route(s)

```routes
/telegram/channel/telegram
```

## 新 RSS 路由检查表 / New RSS Route Checklist
  
- [X] 新的路由 New Route
  - [X] 跟随 [v2 路由规范](https://docs.rsshub.app/joinus/script-standard.html) Follows [v2 Script Standard](https://docs.rsshub.app/en/joinus/script-standard.html)
- [ ] 文档说明 Documentation
  - [ ] 中文文档 CN
  - [X] 英文文档 EN
- [ ] 全文获取 fulltext
  - [ ] 使用缓存 Use Cache
- [ ] 反爬/频率限制 anti-bot or rate limit?
  - [ ] 如果有, 是否有对应的措施? If yes, do your code reflect this sign?
- [X] [日期和时间](https://docs.rsshub.app/joinus/pub-date.html) [date and time](https://docs.rsshub.app/en/joinus/pub-date.html)
  - [X] 可以解析 Parsed
  - [ ] 时区调整 Correct TimeZone
- [X] 添加了新的包 New package added
- [ ] `Puppeteer`

## 说明 / Note

It is impossible to watch videos longer than 20-30 seconds on the web t.me, as well as download attachments using the current web-based `telegram` route.

This PR uses client-level connection when possible, making full channel content available over RSS.

This does not replace the existing web channel parsing, but activates when `TELEGRAM_SESSION` variable is available. It will proxy downloadable content, including streaming videos.

### CODE REVIEW

File: lib/v2/tg/router.js [hunk 1]
```diff
@@ -0,0 +1,4 @@
+module.exports = function (router) {
```
Comment (TonyRL) [inline lib/v2/tg/router.js:1]: https://docs.rsshub.app/joinus/advanced/script-standard

- Do not create variations of the same namespace. Use `telegram`, not `tg` or `t`.

File: lib/v2/tg/channel.js [hunk 2]
```diff
@@ -0,0 +1,170 @@
+const wait = require('@/utils/wait');
+const { client, decodeMedia, getFilename, getMediaLink, streamDocument, streamThumbnail } = require('./client');
+const bigInt = require('telegram/Helpers').returnBigInt;
+const core_router = require('@/core_router');
```
Comment (TonyRL) [inline lib/v2/tg/channel.js:4]: Do not mount on core router.

File: lib/v2/tg/client.js [hunk 3]
```diff
@@ -0,0 +1,194 @@
+const readline = require('node:readline/promises');
+const { TelegramClient } = require('telegram');
+const { StringSession } = require('telegram/sessions');
+const { Api } = require('telegram/tl/api');
+
+const config = require('../../config').value.tg;
+
+const apiId = config.apiId ?? 21834366;
+const apiHash = config.apiHash ?? 'd9601f8a2762a03e910a8cfc2ba79004';
```
Comment (TonyRL) [inline lib/v2/tg/client.js:9]: Do not provision your own api id here. You may include the [official one](https://github.com/DrKLO/Telegram/blob/master/TMessagesProj/src/main/java/org/telegram/messenger/BuildVars.java).

File: lib/v2/tg/channel.js [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+            if (range.length === 1) {
+                // console.log(`${ctx.method} ${ctx.req.url} Range: ${ctx.get('Range')}`);
+                ctx.status = 206; // partial content
+                const [offset, limit] = range[0];
+                ctx.set('Content-Length', limit - offset + 1);
+                ctx.set('Content-Range', `bytes ${offset}-${limit}/${media.document.size}`);
+
+                const stream = streamDocument(media.document, '', offset, limit);
+                let counter = 0;
+                // eslint-disable-next-line no-await-in-loop
```
Comment (TonyRL) [inline lib/v2/tg/channel.js:74]:
```suggestion
```

File: lib/config.js [hunk 5]
```diff
@@ -283,6 +283,12 @@ const calculateValue = () => {
         telegram: {
             token: envs.TELEGRAM_TOKEN,
         },
+        tg: {
+            session: envs.TG_SESSION,
+            apiId: envs.TG_API_ID,
+            apiHash: envs.TG_API_HASH,
+            maxConcurrentDownloads: envs.TG_MAX_CONCURRENT_DOWNLOADS
```
Comment (TonyRL) [inline lib/config.js:290]: Missing documentation on what these value means and how to obtain them.

**Review by TonyRL — COMMENTED** [review]
There're a spam of `RPCError: 400: LIMIT_INVALID (caused by upload.GetFile)`

**Comment by synchrone** [comment]
@TonyRL can you please share the url which triggered RPCError: 400: LIMIT_INVALID (caused by upload.GetFile) ?

**Comment by synchrone** [comment]
@TonyRL @DIYgod can I please get a review? I've been using this personally since August, it seems pretty stable now.

**Comment by synchrone** [comment]
@TonyRL I found the reason of LIMIT_INVALID errors and fixed it. Also pushed html unparse() fixes to get proper formatting.

File: lib/v2/telegram/router.js [hunk 6]
```diff
@@ -1,5 +1,20 @@
+const config = require('@/config').value;
+const core_router = require('@/core_router');
+const tglibChannel = require('./tglib/channel');
+const webChannel = require('./channel');
```
Comment (TonyRL) [inline lib/v2/telegram/router.js:4]:
```suggestion
const coreRouter = require('@/core_router');
```

File: lib/v2/telegram/router.js [hunk 7]
```diff
@@ -1,5 +1,20 @@
...
+    if (config.telegram.session) {
+        // core_router does not cache, which is necessary for video streaming
+        core_router.get('/telegram/channel/:username/:media(.+)', tglibChannel.getMedia);
+    }
+
+    router.get('/channel/:username/:routeParams?', (ctx) => {
+        // tglib impl does not support routeParams yet
+        const useWeb = ctx.params.routeParams || !config.telegram.session;
+        return useWeb ? webChannel(ctx) : tglibChannel(ctx);
+    });
```
Comment (TonyRL) [inline lib/v2/telegram/router.js:16]:
```suggestion
    if (config.telegram.session) {
        // core_router does not cache, which is necessary for video streaming
        coreRouter.get('/telegram/channel/:username/:media(.+)', require('./tglib/channel').getMedia);
    }

    router.get('/channel/:username/:routeParams?', (ctx) => {
        // tglib impl does not support routeParams yet
        const useWeb = ctx.params.routeParams || !config.telegram.session;
        return useWeb ? require('./channel')(ctx) : require('./tglib/channel')(ctx);
    });
```

In this way, gramjs won't [start with RSSHub](https://github.com/DIYgod/RSSHub/actions/runs/7031510001/job/19133265093#step:12:12)

**Comment by synchrone** [comment]
:tada: thank you so much @TonyRL!

**Pull request merged**